### PR TITLE
New: OpenComputers Conduits

### PR DESCRIPTION
**Pull request by HenryLoenwind**
* Open for code review, there must still be a million bugs left...
* OpenComputers conduits ~~based on "one node per block". Later optimizations may reduce this to "one node per outside connection" or "one node per network". But step by step...~~ Now uses one Node per network and channel.
* Also fixed conduit positions for gas and me, so they are always in the same position in regard to the base conduits
* Adds new render option for conduits: Not mirroring the texture. Needed for animated, moving textured
* Textures are based on OCs CC0 licenced textures and are also CC0 licenced
* OC cables support different colors that don't connect to each other. That feature is hardcoded to their cables, so conduits can always connect to all cables
* Multiple channels ~~may be added later. Once the number of nodes has been optimized, otherwise we'd have 16 nodes per block...~~ have been added.

### CODE REVIEW

**Comment by Vexatos** [comment]
:+1:

I guess this should close #725 when merged.

**Comment by Vexatos** [comment]
[So far everything seems to be working](http://gfycat.com/CalmMelodicDeinonychus)... Good job! :+1: 

I would probably try and find a better name than "OC Conduit", something like "Data Conduit" would fit much better...

**Comment by fnuecke** [comment]
Awesome! Just did some testing with the branch, too, and it looks like it's working great :-)

Even helped me find a minor bug in OC (pretty much purely visual, leading to the conduits connecting to the front of server racks).

Random note from debugging that: it seems `canConnectToExternal` only ever gets called on the serverside, and then that info is getting sent to the client manually? In that case the `.isRemote` check in there would be superfluous.

**Comment by HenryLoenwind** [comment]
I did a call stack analysis on canConnectToExternal() and there was plenty of client-side stuff in it. Try opening the conduit core node (shift-left click with a yeta wrench on the core), I think there's one of theose calls in there. All in all, eio code isn't very clean in what code can be called where.

**Comment by fnuecke** [comment]
Ahh, I see, nevermind that then :-)

**Comment by HenryLoenwind** [comment]
Update: This new version only uses one Node per conduit network...and channel. It implements 16 color coded channels, just like the item and insulated redstone conduits.

**Comment by fnuecke** [comment]
FWIW, I gave the branch a good bit of testing, and everything seems to work like it should! :-)

**Comment by tterrag1098** [comment]
I need to pull and test this, will get to it soon

File: src/main/java/crazypants/enderio/conduit/AbstractConduit.java [hunk 1]
```diff
@@ -639,8 +642,16 @@ protected boolean renderStub(ForgeDirection dir) {
 		  return new GasConduitNetwork();
 	  } else if(IMEConduit.class.isAssignableFrom(type)) {
 		  return new MEConduitNetwork();	
+    } else if (IOCConduit.class.isAssignableFrom(type)) {
+      return new OCConduitNetwork();
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/conduit/AbstractConduit.java]: Indentation is all over the place here

File: src/main/java/crazypants/enderio/conduit/ConduitDisplayMode.java [hunk 2]
```diff
@@ -15,6 +16,7 @@
   ITEM,
   GAS, 
   ME,
+ OC,
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/conduit/ConduitDisplayMode.java]: And here

File: src/main/java/crazypants/enderio/conduit/AbstractConduit.java (hunk 1 above)
Comment (HenryLoenwind) [inline src/main/java/crazypants/enderio/conduit/AbstractConduit.java]: Mine is fine, the rest has tabs

File: src/main/java/crazypants/enderio/conduit/ConduitDisplayMode.java (hunk 2 above)
Comment (HenryLoenwind) [inline src/main/java/crazypants/enderio/conduit/ConduitDisplayMode.java]: that's the format.xml, not me

File: src/main/java/crazypants/enderio/conduit/IConduit.java [hunk 3]
```diff
@@ -137,4 +137,10 @@
   
   public AbstractConduitNetwork<?, ?> createNetworkForType();
 
+  /**
+   * Should the texture of the conduit connectors be mirrored around the conduit
+   * node?
+   */
+  boolean getMirrorTexture();
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/conduit/IConduit.java]: I don't really like the use of "get" for boolean states

File: src/main/java/crazypants/enderio/conduit/geom/Offset.java [hunk 4]
```diff
@@ -25,7 +25,9 @@
   SOUTH_DOWN(ForgeDirection.SOUTH, ForgeDirection.DOWN),
   NORTH_DOWN(ForgeDirection.NORTH, ForgeDirection.DOWN),
   NORTH_EAST(ForgeDirection.NORTH, ForgeDirection.EAST), 
-  NORTH_WEST(ForgeDirection.NORTH, ForgeDirection.WEST);
+ NORTH_WEST(ForgeDirection.NORTH, ForgeDirection.WEST), //
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/conduit/geom/Offset.java]: One more indent problem

File: src/main/java/crazypants/enderio/conduit/packet/ConTypeEnum.java [hunk 5]
```diff
@@ -15,7 +16,7 @@
   ITEM(IItemConduit.class),
   REDSTONE(IRedstoneConduit.class),
   GAS(IGasConduit.class),
-  ME(IMEConduit.class);
+ ME(IMEConduit.class), OC(IOCConduit.class);
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/conduit/packet/ConTypeEnum.java]: 2 space indent, and each constant on a new line

File: src/main/java/crazypants/enderio/gui/IconEIO.java [hunk 6]
```diff
@@ -37,6 +37,8 @@
   WRENCH_OVERLAY_ME(32, 128),
   WRENCH_OVERLAY_ME_OFF(32 + 48, 128),
 
+  WRENCH_OVERLAY_OC(96, 112), WRENCH_OVERLAY_OC_OFF(96, 128),
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/gui/IconEIO.java]: New constant on next line

File: src/main/java/crazypants/enderio/conduit/AbstractConduit.java (hunk 1 above)
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/conduit/AbstractConduit.java]: Just reformat the whole thing then, this is more of that code from the broken PR -.-

**Comment by SquidDev** [comment]
When this was announced on Reddit a couple of months ago there was [some interest](https://www.reddit.com/r/feedthebeast/comments/3ks4fu/enderio_is_getting_a_new_conduit_type_one_that/cv01c55) in adding support for ComputerCraft via [CCTweaks](https://github.com/SquidDev-CC/CC-Tweaks). Would it be worth me putting a PR together on this, or seeing as it isn't supported by CC official is there no point?

**Comment by tterrag1098** [comment]
Maybe, but it'd probably have to wait for 2.4.X

**Comment by HenryLoenwind** [comment]
I started on the CCTweaks conduits but put the work on hold until the OC conduits went live. Took a bit longer than expected and as tt said, we are now in the stabilization phase for 2.3.

**Comment by JoeJoeTV** [comment]
> I started on the CCTweaks conduits but put the work on hold until the OC conduits went live. Took a bit longer than expected and as tt said, we are now in the stabilization phase for 2.3.

Are you still working on the CC conduits?

**Comment by tyler489** [comment]
Holy necro

**Pull request merged**